ingest.py

```diff
@@ -0,0 +1,310 @@
+
+def is_valid_first_segment(first_segment):
+    return len(first_segment) == 4 and first_segment[0] in valid_first_segment_first_char and first_segment[1:] in valid_first_segment_other_chars
+
+def is_valid_second_segment(date_segment):
+    try:
+        datetime.strptime(date_segment, "%Y-%m-%d")
+        return True
+    except ValueError:
+        return False
+
+def is_valid_id(id_segment):
+    valid_initials = ''.join(valid_id_initial_chars)
+    id_pattern = rf"^\d{{7}}$|^[{valid_initials}]\d{{6}}$"
+    parts = id_segment.split('-')
+    return all(re.fullmatch(id_pattern, part) for part in parts)
+
+def is_valid_freetext_segment(freetext_segment):
+    if freetext_segment.startswith("s-"):
+        return False
+    return re.fullmatch(r"^[a-z0-9][a-z0-9-]*$", freetext_segment) is not None
+
+def is_valid_suffix_segment(suffix_segment):
+    if not suffix_segment.startswith("s-"):
+        return False
+    parts = suffix_segment[2:].split("-")
+    for sub in parts:
+        if sub in valid_suffixes:
+            continue
+        elif re.fullmatch(r"\d{3}", sub):
+            continue
+        else:
+            return False
+    return True
+
+def is_valid_filename(file_name):
+    base_file_name, _ = os.path.splitext(file_name)
+    segments = base_file_name.split('_')
+
+    if len(segments) < 2:
+        logging.warning(f"{file_name}: Too few segments")
+        return False, True
+
+    if not is_valid_first_segment(segments[0]):
+        logging.warning(f"{file_name}: Invalid first segment")
+        return False, True
+
+    idx = 1
+
+    # Optional ID segment
+    if len(segments) > idx and is_valid_id(segments[idx]):
+        idx += 1
+
+    # Mandatory date segment
+    if len(segments) <= idx or not is_valid_second_segment(segments[idx]):
+        logging.warning(f"{file_name}: Missing or invalid date segment")
+        return False, True
+    idx += 1
+
+    # Optional freetext
+    if len(segments) > idx and is_valid_freetext_segment(segments[idx]):
+        idx += 1
+
+    # Optional suffix
+    if len(segments) > idx and is_valid_suffix_segment(segments[idx]):
+        idx += 1
+
+    if len(segments) > idx:
+        logging.warning(f"{file_name}: Too many segments")
+        return False, True
+
+    return True, False
+
+def get_metadata_tags(file_path):
+    try:
+        # Run exiftool to get JSON output
+        result = subprocess.run(
+            ['exiftool', '-j', file_path],
+            capture_output=True,
+            text=True,
+            check=True
+        )
+        metadata_list = json.loads(result.stdout)
+        if metadata_list:
+            return metadata_list[0]  # first dict in list contains metadata
+        return {}
+    except Exception as e:
+        logging.error(f"Failed to get metadata for {file_path}: {e}")
+        return {}
+
+def missing_required_metadata(metadata):
+    missing = []
+    for tag in required_metadata_tags:
+        if tag not in metadata or not metadata[tag]:
```

Comment:
```suggestion
        if tag not in metadata or not metadata[tag]:
```
What is the `metadata[tag]` for? Doctest would help again.